contents/javascript/Javascript-framework-React.md

```diff
@@ -0,0 +1,152 @@
+
+<box type="info">
+    Notice the HTML-like syntax within the JavaScript method <code>render()</code>. This is <a href="https://reactjs.org/docs/introducing-jsx.html">JSX</a>, and it is a JavaScript syntax extension developed by the React team. React <a href="https://reactjs.org/docs/introducing-jsx.html#why-jsx">recommends</a> developers to use JSX, but it is optional to use it.
+</box>
+
+In the example above, we reused the markup for each heading using an `ExampleComponent`. A React **component** is just a JavaScript function that accepts arbitrary input (called `props`) and returns elements describing what should appear on the screen. 
+
+Components can be especially useful when your UI is complex. With components, you can not only *reuse* code, but also *divide* your UI into encapsulated units that manage their own data and logic. 
+
+For example, if you have a todo list application, it could have a `Form` component to handle adding new items and a `TodoList` component to display all the items. The `TodoList` component could further be composed of multiple `TodoListItem` components, one for each item in the todo list. 
+
+![Building a Todo List application using React](javascript-framework-react-imgs/ReactComponentsExample.png "Building a Todo List application using React")
+
+You can read more about components and props in the React [documentation](https://reactjs.org/docs/components-and-props.html). 
+
+#### Benefit: Better Performance
+
+Web applications can have a lot of user interaction and data updates, which results in changes being made to the DOM. Adding and removing DOM nodes isn't slow. The performance-bottleneck that arises due to DOM manipulation is because the browser needs to [reflow and repaint](https://www.phpied.com/rendering-repaint-reflowrelayout-restyle/) the UI every time the DOM is changed.
+
+React minimizes this update time by using a virtual DOM. The virtual DOM is a JavaScript object that is kept in the memory of your application.
+
+![React's Virtual DOM](javascript-framework-react-imgs/ReactVirtualDOM.png "React's Virtual DOM")
+
+Any updates to the UI will first be made to the virtual DOM. Then, React will compare the virtual DOM with the actual DOM using a [diffing algorithm](https://reactjs.org/docs/reconciliation.html#the-diffing-algorithm). 
+Finally, React updates the actual DOM only in places it differs with the virtual DOM. It batches multiple changes together and updates the actual DOM in one go, minimizing update time. 
+
+#### Other Advantages of React
+
+Besides the three main benefits explained above, React has the following advantages:
+ - React only allows data to flow downwards (one-way data binding), which makes your application [easier to debug](https://flaviocopes.com/react-unidirectional-data-flow/).
```

Comment:
the link doesn't seem relevant to the hyperlink text 🤔

Reply:
It has the benefits of one-way data binding listed out wrt debugging after the introduction actually.

Reply:
Yep, it states the `benefits`, whereas the hyperlink text is `easier to debug`, which I expected to explain  the how and why in more detail (_maybe just me_). 
Maybe you can paraphrase and add `easier to debug, as you know what is coming from where` in your article?

Reply:
Ah, okay. I can paraphrase.